### PR TITLE
fix: ensure packets are emitted in order #24

### LINKED ISSUE #24: [bug] Memory leak, missing packets
On high loads, the mqemitter miss to emit some packets.

The leak is [here](https://github.com/mcollina/mqemitter-mongodb/blob/master/mqemitter-mongodb.js#L183-L206) 

I'm unable to find a fix to this. 

Reference:
https://github.com/moscajs/aedes/issues/533

### PR DESCRIPTION
**Pull request by robertsLando**
Fixes #24

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
lgtm

**Comment by robertsLando** [comment]
@mcollina now test fails, as expected

**Comment by robertsLando** [comment]
@mcollina I have a question, If I call multiple emits, should messages be reported in order even  if I don't wait for emit callback before each emit?

**Comment by mcollina** [comment]
> I have a question, If I call multiple emits, should messages be reported in order even if I don't wait for emit callback before each emit?

In theory yes, I would expect such a behavior.

**Comment by robertsLando** [comment]
@mcollina Ok so the test was good in the previous format, if I wait for callbacks it works but if not it doesn't, and that is the case of aedes, it is emitting callbacks in queue order but sometimes it emits one before another and that's why aedes is missing some packets, because when this happens the packet is deduped and discarded. I think the leak may be not in the code we pointed out.

Before calling a callback we should ensure that all previous callbacks have been called

**Comment by mcollina** [comment]
> Ok so the test was good in the previous format, if I wait for callbacks it works but if not it doesn't, and that is the case of aedes, it is emitting callbacks in queue order but sometimes it emits one before another and that's why aedes is missing some packets, because when this happens the packet is deduped and discarded. I think the leak may be not in the code we pointed out.

I do not understand how this could happen.

**Comment by robertsLando** [comment]
@mcollina Check the test now

Also check this line: https://github.com/mcollina/mqemitter-mongodb/blob/master/mqemitter-mongodb.js#L180

If this is true are we 100% sure that there is no other packets waiting to be emittet with an id < then last id?

**Comment by robertsLando** [comment]
@mcollina Check tests now, I think the problem is here: https://github.com/mcollina/mqemitter-mongodb/blob/master/mqemitter-mongodb.js#L137

Here we emit packets everytime we receive a packet from the stream, but that stream outputs packet when they are successfully stored in mongodb but not in the order [`insertOne`](https://github.com/mcollina/mqemitter-mongodb/blob/master/mqemitter-mongodb.js#L166) is called

Also I think that the test I have added should be added to abstract tests of mqemitter

**Comment by robertsLando** [comment]
I have opened a PR on mqemitter: https://github.com/mcollina/mqemitter/pull/34

**Comment by robertsLando** [comment]
@mcollina Maybe this approach could be performance killer (or maybe not) but with this fix the leak problem is fixed!

**Comment by robertsLando** [comment]
The only problem I have found here is that when the 'leak' happens I have to immediatly call `cb`  here:

https://github.com/mcollina/mqemitter-mongodb/pull/26/files#diff-614d4886bb4d8e8781639ee81936918cR146

Do you have any suggestion? Without that the process cannot continue but with this approach the cb passed to the emitted event is useless

**Review by mcollina — COMMENTED** [review]
I would recommend to use fastq instead.

**Comment by robertsLando** [comment]
> I would recommend to use fastq instead.

I dunno how that could be used here, how can I do a find in that queue? Also isn't that queue used for functions? I'm using it to store packets

**Comment by robertsLando** [comment]
Just thinking that this may not work in a cluster env and could still have some leaks, let me check if it can be fixed using writeConcern

**Comment by robertsLando** [comment]
I have tried to use 

`    this._collection.insertOne(obj, { w: 0, j: false, wtimeout: 100 }, function (err, res) { `

without success, on high loads it still happens that two consecutive inserts are not stored in consecutive order (and emitted in order in the stream)

So ATM this fixes the problem but it will persist when using it in clusters environments

**Comment by mcollina** [comment]
Have you tried using `w: 1`?

**Comment by robertsLando** [comment]
@mcollina That's the default value. But even with/without that nothing changes. Sincerly I don't know if there is a solution to this (in clusters)

I cannot find nothing about ways to ensure that documents are written in the same order the insert is called

**Comment by mcollina** [comment]
I don't think there is a problem in cluster because the dedupe logic in aedes is per Node.js single process.

**Comment by robertsLando** [comment]
@mcollina You right, missed this :confounded: Ok so let me check why it's failing now

**Comment by robertsLando** [comment]
@mcollina My only perplexity is about the callback in `process` (check the TODO)

**Comment by robertsLando** [comment]
@mcollina Now should be ready, I have also fixed the callback logic now :tada:

File: mqemitter-mongodb.js [hunk 1]
```diff
@@ -149,6 +150,37 @@ function MQEmitterMongoDB (opts) {
 
 inherits(MQEmitterMongoDB, MQEmitter)
 
+MQEmitterMongoDB.prototype._findNext = function(id) {
+  for (var i = 0, len = this._queue.length; i < len; i++) {
+    if(this._queue[i]._id.toString() === id) {
```
Comment (mcollina) [inline mqemitter-mongodb.js]: these are a lot of .toString() operations for ._id. Maybe we can store it in already in string form?

**Review by mcollina — APPROVED** [review]
lgtm

File: mqemitter-mongodb.js (hunk 1 above)
Comment (robertsLando) [reply mqemitter-mongodb.js]: I know but it must be an ObjectId when I assign it to lastId to make the checks in emit, the alternative would be to store it in string format and convert it to objectId when I assign it to lastId but I don't think there is any optimization in this as there would be lot `new ObjectId(id)` in place of `.toString()`. An alternative would be to add another property like `_idString` to the object and use that to compare

**Comment by robertsLando** [comment]
@mcollina Would you mind also to add a new version of mqemitter so I can update it here and remove the test?

File: mqemitter-mongodb.js (hunk 1 above)
Comment (robertsLando) [reply mqemitter-mongodb.js]: Also, in 99.99% of cases the _findNext will always return 0 so it will do only one check, the 0.01% will return 1

**Comment by robertsLando** [comment]
I'm thinking again to https://github.com/mcollina/mqemitter-mongodb/pull/26#issuecomment-684829044

We have 2 aedes instances: 

1. aedes1-mq1
2. aedes2-mq2

- mq1 inserts 2 packets (A B), them are received by the stream as (B A) but the queue fix this and emit them to aedes1 as (A B)
- mq2 stream receives it in the wrong order (B A) as it hasn't the packet queue and doesn't know the order, so it emits them in the wrong order to aedes2 (B A)

The bug in this case is still present as different instances may receive packets in a different order. (This also explains why tests were flaky in [aedes-tests](https://github.com/moscajs/aedes-tests) with mongodb)

PS: I have already tested this with aedes-test clusters and it works, sure better then before but still not perfect

**Comment by robertsLando** [comment]
The only solution to this could be a shared queue, so another collection that keeps track of the emitted packet order, this would slow down us a lot as the read/writes in mongo will be double

**Comment by robertsLando** [comment]
@mcollina ping

File: mqemitter-mongodb.js (hunk 1 above)
Comment (mcollina) [reply mqemitter-mongodb.js]: Give that we store it in an object, we can just store it in both formats.

Note that we use it here and then in `_emitPacket` as well.

File: test.js [hunk 2]
```diff
@@ -3,12 +3,14 @@
...
 var MongoClient = mongodb.MongoClient
 var MongoEmitter = require('./')
-var test = require('tape').test
+var { test } = require('tape')
 var abstractTests = require('mqemitter/abstractTest.js')
 var clean = require('mongo-clean')
 var dbname = 'mqemitter-test'
 var url = 'mongodb://127.0.0.1/' + dbname
 
+console.time('test')
```
Comment (mcollina) [inline test.js]: this is not needed

**Review by mcollina — COMMENTED** [review]
I have also released mqemitter@4.4.0.

**Comment by robertsLando** [comment]
@mcollina done. Anyway as I explain in https://github.com/mcollina/mqemitter-mongodb/pull/26#issuecomment-684933211 I think the bug will persist in a cluster env

**Comment by robertsLando** [comment]
@mcollina I have found the final solution :heart_eyes: , completely refactored all previous changes I have now used orderedBulkOperations to ensure the write order, performances are much better now and also this works in clusters :tada: :rocket:

**Review by mcollina — APPROVED** [review]
lgtm!

**Comment by robertsLando** [comment]
Feel free to merge and release Matteo I'm done here 😊

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**